Add hoomd.version.floating_point_precision and summarize features (#1211)
glotzerlab/hoomd-blue: sphinx-doc/features.rst

.. Copyright (c) 2009-2022 The Regents of the University of Michigan.
.. Part of HOOMD-blue, released under the BSD 3-Clause License.

Features
========

Hard particle Monte Carlo
-------------------------

HOOMD-blue can perform simulations of hard particles using the Monte Carlo method (`hpmc`). Hard

> particles are defined by their shape and the `HPMC integrator <hpmc.integrate>` supports

Reviewer: ```suggestion
particles are defined by their shape, and the `HPMC integrator <hpmc.integrate>` supports
```

(`md.minimize`). Unless otherwise stated in the documentation, all integration methods integrate
both translational and rotational degrees of freedom. Some integration methods support manifold
constraints (`md.methods.rattle`). HOOMD-blue provides a number of pair potentials (`md.pair`)
including pair potentials that depend on particle orientation (`md.pair.aniso`) and many body
potentials (`md.many_body`). HOOMD-blue also provides bond potentials and distance constraints
commonly used in atomistic and coarse-grained force fields (`md.angle`, `md.bond`,
`md.constrain.Distance`, `md.dihedral`, `md.improper`, `md.special_pair`) and can model rigid bodies
(`md.constrain.Rigid`). External fields `md.external.field` apply potentials based only on the
particle's position and orientation, including walls (`md.external.wall`) to confine particles in a
specific region of space. `md.long_range` provides long ranged interactions, including the PPPM
method for electrostatics. HOOMD-blue enables active matter simulations with `md.force.Active` and
`md.update.ActiveRotationalDiffusion`. At runtime, `hoomd.version.md_built` indicates whether the
build supports MD simulations.

> md.update

Reviewer: typo?

Author: Yes.

extend HOOMD-blue with functionality implemented in Python code by subclassing `trigger.Trigger`,
`variant.Variant`, `hoomd.update.CustomUpdater`, `hoomd.write.CustomWriter`,
`hoomd.tune.CustomTuner`, or by using the HOOMD-blue API in combination with other Python packages
to implement methods that couple simulation, analysis, and multiple simulations (such as umbrella
sampling).

.. seealso::

    Tutorial: :doc:`tutorial/04-Custom-Actions-In-Python/00-index`

CPU and GPU devices
-------------------

HOOMD-blue can execute simulations on CPUs or GPUs. Typical simulations run more efficiently on
GPUs for system sizes larger than a few thousand particles, although this strongly depends on the
details of the simulation. The provided binaries support NVIDIA GPUs. Build from source to enable
preliminary support for AMD GPUs. CPU support is always enabled. GPU support must be enabled at
compile time with the ``ENABLE_GPU`` CMake option (see :doc:`building`). Select the device to use at
run time with the `device <hoomd.device>` module. Unless otherwise stated in the documentation,
**all** operations and methods support GPU execution. At runtime, `hoomd.version.gpu_enabled` indicates

> whether the build supports run time compilation.

Reviewer: Does run time compilation mean JIT code?

Or does it mean "able to run on GPU"?

```suggestion
whether the build supports running on the GPU.
```

Author: Fixed.

MPI
---

HOOMD-blue can use the message passing interface (MPI) to execute simulations in less time using
more than one CPU core or GPU. Unless otherwise stated in the documentation, **all** operations and
methods support MPI parallel execution. MPI support is optional, requires a compatible MPI library,
and must be enabled at compile time with the ``ENABLE_MPI`` CMake option (see :doc:`building`).
At runtime, `hoomd.version.mpi_enabled` indicates whether the build supports MPI.

.. seealso::

    Tutorial: :doc:`tutorial/03-Parallel-Simulations-With-MPI/00-index`

Threading
---------

Some operations in HOOMD-blue can use multiple CPU threads in a single process. Control this with

> the `device.Device.num_cpu_threads` property. In this release, threading support in HOOMD-blue is

Reviewer: "in this release"

Is there a way to set a reminder to update this when that changes? Maybe just a code comment

Author: We just need to be diligent. There is no direct connection between what is written in the code and the docs. This may never change. At best I would hope that we remember to update the local docs for each operation when we change them. But the overview docs are the most likely to get out of date.

Reviewer: maybe just tag with like "as of 3.0" so readers know when it was update?
```suggestion
the `device.Device.num_cpu_threads` property. As of version 3.0, threading support in HOOMD-blue is
```

very limited and only applies to implicit depletants in `hpmc.integrate.HPMCIntegrator`, and
`hpmc.pair.user.CPPPotentialUnion`. Threading must must be enabled at compile time with the
``ENABLE_TBB`` CMake option (see :doc:`building`). At runtime, `hoomd.version.tbb_enabled` indicates

> whether the build supports MPI.

Reviewer: Reading this I'm confused about the discrepancy between the flag saying TBB and the description saying MPI again.

Author: Right, copy and paste bug. Fixed.

Run time compilation
--------------------

> Some operations allow the user to provide arbitrary C++ code which HOOMD-blue compiles at run time

Reviewer: ```suggestion
Some operations allow the user to provide arbitrary C++ code that HOOMD-blue compiles at run time
```

and executes during the simulation. `hpmc.pair.user` and `hpmc.external.user` enable users to apply

> arbiratray pair and external potentials to particles in HPMC simulations. `hpmc.pair.user` is

Reviewer: ```suggestion
arbitrary pair and external potentials to particles in HPMC simulations. `hpmc.pair.user`
```

> supports both CPUs and NVIDIA GPUs while `hpmc.external.user` is only supports CPUs. Run time

Reviewer: ```suggestion
supports both CPUs and NVIDIA GPUs while `hpmc.external.user` only supports CPUs. Run time
```

compilation must be enabled at compile time with the ``ENABLE_LLVM`` CMake option (see
:doc:`building`). At runtime, `hoomd.version.llvm_enabled` indicates whether the build supports run
time compilation.

Mixed precision
---------------

HOOMD-blue performs computations with mixed floating point precision. There is a **high precision**

> type and a **reduced precision** type. All particle properties are stored in the **high precision**

Reviewer: I would not bold the high/reduced precision terms after they are first mentioned. It provides too much emphasis. Perhaps use a code syntax style?

Author: I just removed the bold.

> type and most operations also perform all computations with **high precision**. Operations that to

Reviewer: ```suggestion
type, and most operations also perform all computations with **high precision**. Operations that do
```
